### ISSUE TITLE
Writing code generic over client type?

### ISSUE THREAD

**Issue by netvl** [async-email/async-imap#18]
Currently `Client` is parametrized over the `Read + Write + fmt::Debug` types. Unfortunately, because this is a requirement of three non-marker traits, it does not seem possible to write code which is agnostic to the specific kind of the underlying stream.

My use case is that I need to write some code which works with IMAP servers, but depending on configuration those servers can be TLS or non-TLS. As of now, it does not seem possible to handle such servers in a uniform way.

I'm frankly not sure how, but it would be great if the library supported some kind of a way to write code which does not depend on the connection details of a particular client and would allow working with any of them. I guess the most straightforward way to do it would be to extract the public interface of the `Client` and `Session` structs to traits, but given that traits can't have async method now, it might be hard to do...

**Comment by dignifiedquire** [comment]
I agree this is an issue, I have not found a great solution that I like yet, but I definitely experienced this, which is why this file (https://github.com/deltachat/deltachat-core-rust/blob/master/src/imap_client.rs) exists. 



> but given that traits can't have async method now, it might be hard to do...

[async-trait](https://docs.rs/async-trait/0.1.22/async_trait/) works quite well, so that might be possible.

**Comment by dignifiedquire** [comment]
we could take some inspiriation from here https://docs.rs/async-tungstenite/0.3.1/async_tungstenite/stream/enum.Stream.html (haven't looked in detail yet)

**Comment by bnjbvr** [comment]
While trying to replace `imap` with `async-imap` for some customer project, I've come up with the following: https://gist.github.com/bnjbvr/9c47d5cce34fa22b0f71601bd3d4f9dc

This reproduces `imap`'s `ConnectionMode` enum, so it's possible to connect with either of TLS, STARTTLS and plaintext mode. I've only tested it against my TLS server, but from reading the documentation it _should? may?_ work with STARTTLS and plaintext mode — extra testing would be appreciated, of course :)

Would that be of interest as a contribution to the project, assuming it can be tested automatically?
